File: university-backend/src/main/java/com/psdku/lmj/university_backend/service/AnnouncementService.java
/*
 * Click nbfs://nbhost/SystemFileSystem/Templates/Licenses/license-default.txt to change this license
 * Click nbfs://nbhost/SystemFileSystem/Templates/Classes/Class.java to edit this template
 */

/*
 * Announcement Service
 */
package com.psdku.lmj.university_backend.service;

import com.psdku.lmj.university_backend.model.Announcement;
import com.psdku.lmj.university_backend.model.Student;
import com.psdku.lmj.university_backend.repository.AnnouncementRepository;
import com.psdku.lmj.university_backend.repository.StudentRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.Optional;

/**
 * Announcement Service
 * 
 * Business logic for managing announcements
 */
@Service
public class AnnouncementService {
    
    @Autowired
    private AnnouncementRepository announcementRepository;
    
    @Autowired
    private StudentRepository studentRepository;
    
    /**
     * Get all announcements for a student
     */
    public List<Announcement> getStudentAnnouncements(Long studentId) {
        return announcementRepository.findByStudentId(studentId);
    }
    
    /**
     * Get recent announcements for a student (default: 5 most recent)
     */
    public List<Announcement> getRecentAnnouncements(Long studentId) {
        return announcementRepository.findRecentByStudentId(studentId, 5);
    }
    
    /**
     * Get recent announcements with custom limit
     */
    public List<Announcement> getRecentAnnouncements(Long studentId, int limit) {
        return announcementRepository.findRecentByStudentId(studentId, limit);
    }
    
    /**
     * Get important announcements for a student
     */
    public List<Announcement> getImportantAnnouncements(Long studentId) {
        return announcementRepository.findImportantByStudentId(studentId);
    }
    
    /**
     * Get specific announcement by ID
     */
    public Optional<Announcement> getAnnouncementById(Long id) {
        return announcementRepository.findById(id);
    }
    
    /**
     * Create new announcement for a student
     */
    public Announcement createAnnouncement(Long studentId, Announcement announcement) {
        Optional<Student> studentOptional = studentRepository.findById(studentId);
        
        if (studentOptional.isEmpty()) {
            throw new RuntimeException("Student not found with id: " + studentId);
        }
        
        announcement.setStudent(studentOptional.get());
        return announcementRepository.save(announcement);
    }
    
    /**
     * Update announcement
     */
    public Announcement updateAnnouncement(Long id, Announcement announcementDetails) {
        Announcement announcement = announcementRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Announcement not found with id: " + id));
        
        if (announcementDetails.getTitle() != null) {
            announcement.setTitle(announcementDetails.getTitle());
        }
        if (announcementDetails.getDate() != null) {
            announcement.setDate(announcementDetails.getDate());
        }
        if (announcementDetails.getImportant() != null) {
            announcement.setImportant(announcementDetails.getImportant());
        }
        
        return announcementRepository.save(announcement);
    }
    
    /**
     * Delete announcement
     */
    public void deleteAnnouncement(Long id) {
        Announcement announcement = announcementRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Announcement not found with id: " + id));
        
        announcementRepository.delete(announcement);
    }
    
    /**
     * Get all announcements (admin only)
     */
    public List<Announcement> getAllAnnouncements() {
        return announcementRepository.findAll();
    }
}

File: university-frontend/src/app/services/announcement.service.ts
// src/app/services/announcement.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../environments/environment';

export interface Announcement {
  id?: number;
  title: string;
  date: string;
  important: boolean;
}

@Injectable({
  providedIn: 'root'
})
export class AnnouncementService {
  private apiUrl = `${environment.apiUrl}/announcements`;

  constructor(private http: HttpClient) {}

  /**
   * Get all announcements for a specific student
   */
  getStudentAnnouncements(studentId: number): Observable<Announcement[]> {
    return this.http.get<Announcement[]>(`${this.apiUrl}/student/${studentId}`);
  }

  /**
   * Get recent announcements for a student (default: 5 most recent)
   */
  getRecentAnnouncements(studentId: number, limit: number = 5): Observable<Announcement[]> {
    return this.http.get<Announcement[]>(`${this.apiUrl}/student/${studentId}/recent?limit=${limit}`);
  }

  /**
   * Get important announcements for a student
   */
  getImportantAnnouncements(studentId: number): Observable<Announcement[]> {
    return this.http.get<Announcement[]>(`${this.apiUrl}/student/${studentId}/important`);
  }

  /**
   * Get specific announcement by ID
   */
  getAnnouncementById(id: number): Observable<Announcement> {
    return this.http.get<Announcement>(`${this.apiUrl}/${id}`);
  }

  /**
   * Create new announcement for a student
   */
  createAnnouncement(studentId: number, announcement: Announcement): Observable<Announcement> {
    return this.http.post<Announcement>(`${this.apiUrl}/student/${studentId}`, announcement);
  }

  /**
   * Update announcement information
   */
  updateAnnouncement(id: number, announcement: Partial<Announcement>): Observable<Announcement> {
    return this.http.put<Announcement>(`${this.apiUrl}/${id}`, announcement);
  }

  /**
   * Delete an announcement
   */
  deleteAnnouncement(id: number): Observable<any> {
    return this.http.delete(`${this.apiUrl}/${id}`);
  }

  /**
   * Get all announcements (admin only)
   */
  getAllAnnouncements(): Observable<Announcement[]> {
    return this.http.get<Announcement[]>(this.apiUrl);
  }
}